### PR TITLE
JS: Added support for Array.prototype.[findLastIndex, findLast] ES2022 feature

### PR DESCRIPTION
**Pull request by Napalys**
Added support for ES2023 feature, `Arrray.protype.findLastIndex`, `Arrray.protype. findLast`.

DCA run shows that the run time stays approximately the same. Meta queries revealed new taint for `Arrray.protype.find` calls with lambda, which was expected.

### CODE REVIEW

File: javascript/ql/test/library-tests/Arrays/arrays.js [hunk 1]
```diff
@@ -112,19 +112,19 @@
...
+    const element = list.find((item) => sink(item)); // NOT OK
     sink(element); // NOT OK
   }
 
   {  // Test for findLastIndex function
     const list = ["source"];
-    const element = list.findLastIndex((item) => sink(item)); // NOT OK -- Not caught, currently missing dataflow tracking.
+    const element = list.findLastIndex((item) => sink(item)); // NOT OK
     sink(element); // OK
   }
```
Comment (erik-krogh) [inline javascript/ql/test/library-tests/Arrays/arrays.js:129]: Could you also add a test like: 

```
    const element = source.findLastIndex((item) => sink(item)); // NOT OK - only found with taint-tracking. 
    sink(element); // OK
```

Comment (Napalys) [reply javascript/ql/test/library-tests/Arrays/arrays.js]: Adds test case: f1e95a8a1dee4d68038779011bf3414564080f4f

File: javascript/ql/lib/semmle/javascript/Arrays.qll [hunk 2]
```diff
@@ -492,7 +492,20 @@ private module ArrayLibraries {
...
+  }
+
+  /**
+   * This step models the propagation of data from the array to the callback function's parameter.
+   */
+  private class ArrayCallBackDataTaintStep extends TaintTracking::SharedTaintStep {
+    override predicate step(DataFlow::Node obj, DataFlow::Node element) {
+      exists(DataFlow::MethodCallNode call |
+        call.getMethodName() = ["findLast", "find", "findLastIndex"] and
+        obj = call.getReceiver().getALocalSource() and
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/Arrays.qll:508]:
```suggestion
        obj = call.getReceiver() and
```

I don't think you need `.getALocalSource()` here.  
Try to see if the tests work with this change.

File: javascript/ql/lib/semmle/javascript/Arrays.qll [hunk 3]
```diff
@@ -492,7 +492,20 @@ private module ArrayLibraries {
       exists(DataFlow::MethodCallNode call |
         call.getMethodName() = ["findLast", "find", "findLastIndex"] and
         prop = arrayLikeElement() and
-        obj = call.getReceiver() and
+        obj = call.getReceiver().getALocalSource() and
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/Arrays.qll:495]:
```suggestion
        obj = call.getReceiver() and
```
In the `storeStep` you definitely need `.getALocalSource()`, but I'm not sure you need it here? 
See if this suggestion still passes the tests.

**Review by erik-krogh — COMMENTED** [review]
Two small suggestions to look into.  

Otherwise OK 👍

File: javascript/ql/lib/semmle/javascript/Arrays.qll (hunk 2 above)
Comment (Napalys) [reply javascript/ql/lib/semmle/javascript/Arrays.qll]: 9dbf7d1828b2b2f70cd26c346956ad091993c14f fixes.

File: javascript/ql/lib/semmle/javascript/Arrays.qll (hunk 3 above)
Comment (Napalys) [reply javascript/ql/lib/semmle/javascript/Arrays.qll]: 64c45debdb4c1326f933d100eacac71c7898792d fixes.

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**